sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala

```diff
@@ -755,46 +743,61 @@ class Analyzer(
+    //     - If the table name has 2 parts and the first name part is equal to the global temp view
+    //       database name, resolve the table name to a global temp view.
+    //     - If the first name part matches a registered v2 catalog, look up the table from that
+    //       catalog.
+    //     - Otherwise, look up the table from the current catalog.
     private def lookupTableFromCatalog(
-        tableIdentifier: TableIdentifier,
         u: UnresolvedRelation,
         defaultDatabase: Option[String] = None): LogicalPlan = {
-      val tableIdentWithDb = tableIdentifier.copy(
-        database = tableIdentifier.database.orElse(defaultDatabase))
-      try {
-        v1SessionCatalog.lookupRelation(tableIdentWithDb)
-      } catch {
-        case _: NoSuchTableException | _: NoSuchDatabaseException =>
-          u
+      val nameParts = u.multipartIdentifier
+      assert(nameParts.nonEmpty)
+      val expandedNameParts = defaultDatabase.toSeq ++ nameParts
+      if (expandedNameParts.length == 1) {
+        val tblName = expandedNameParts.head
+        v1SessionCatalog.lookupTempView(tblName).orElse {
+          val tableName = catalogManager.currentNamespace ++ expandedNameParts
+          loadTable(catalogManager.currentCatalog.asTableCatalog, tableName)
+        }.getOrElse(u)
+      } else {
+        val maybeGlobalTempView = if (expandedNameParts.length == 2) {
+          v1SessionCatalog.lookupGlobalTempView(expandedNameParts.head, expandedNameParts.last)
+        } else {
+          None
+        }
+        maybeGlobalTempView.orElse {
+          val CatalogAndIdentifierParts(resolvedCatalog, restNameParts) = expandedNameParts
+          loadTable(resolvedCatalog, restNameParts)
+        }.getOrElse(u)
       }
     }
 
-    // If the database part is specified, and we support running SQL directly on files, and
-    // it's not a temporary view, and the table does not exist, then let's just return the
-    // original UnresolvedRelation. It is possible we are matching a query like "select *
-    // from parquet.`/path/to/query`". The plan will get resolved in the rule `ResolveDataSource`.
-    // Note that we are testing (!db_exists || !table_exists) because the catalog throws
-    // an exception from tableExists if the database does not exist.
-    private def isRunningDirectlyOnFiles(table: TableIdentifier): Boolean = {
-      table.database.isDefined && conf.runSQLonFile && !v1SessionCatalog.isTemporaryTable(table) &&
-        (!v1SessionCatalog.databaseExists(table.database.get)
-          || !v1SessionCatalog.tableExists(table))
+    private def loadTable(tableCatalog: CatalogPlugin, name: Seq[String]): Option[LogicalPlan] = {
+      if (CatalogV2Util.isSessionCatalog(tableCatalog)) {
+        CatalogV2Util.loadTable(tableCatalog, name.asIdentifier).map {
+          case v1Table: V1Table => v1SessionCatalog.createV1Relation(v1Table.v1Table)
+          case v2Table => DataSourceV2Relation.create(v2Table)
```

Comment:
If tableCatalog is session catalog, loadTable only returns V1Table, isn't?

```scala
override def loadTable(ident: Identifier): Table = {
  ...
  V1Table(catalogTable)
}
```

Reply:
Yes, but it will return v2 table after https://github.com/apache/spark/pull/25651

The code was already written in a [future-proof way](https://github.com/apache/spark/pull/26214/files#diff-57b3d87be744b7d79a9beacf8e5e5eb2L2779), so I just keep it.